### PR TITLE
tweak that simplifies Worker startup code

### PR DESCRIPTION
**Pull request by bradygaster**
Do not merge yet. Reviewing with @davidfowl

### CODE REVIEW

**Comment by davidfowl** [comment]
One thing to discuss @halter73 and @eerhardt is that we didn't expose IHostBuilder on the host application builder. I think that will break things like serilog and others that don't have IServiceCollection or IApplicationBuilder extensions methods. It probably also break orleans right @ReubenBond ?

**Comment by halter73** [comment]
> One thing to discuss @halter73 and @eerhardt is that we didn't expose IHostBuilder on the host application builder. I think that will break things like serilog and others that don't have IServiceCollection or IApplicationBuilder extensions methods. It probably also break orleans right @ReubenBond ?

A [HostApplicationBuilder adapter does exist.](https://github.com/dotnet/runtime/blob/9a3626e16efb4fb1ae75d806bf66f0769bd842c1/src/libraries/Microsoft.Extensions.Hosting/src/HostApplicationBuilder.cs#L232-L381) It is exposed through a diagnostic listener for testing and tooling scenarios, but we could not get it approved as a public API.

The problem with exposing an IHostBuilder adapter is that the HostApplicationBuilder cannot do everything an IHostBuilder can after it's constructed. Some things like the ContentRootPath that cannot be changed, so it wouldn't be compatible with every IHostBuilder extension method, but it might be compatible with most. I'm not sure. It's annoying that a lack of support turns into a NotSupportedException at startup, but since this would be new API, maybe that's okay.

During the HostApplicationBuilder API review, we suggested that extension methods that currently target IHostBuilder instead target IConfigurationBuilder and/or IServiceCollection directly. This might mean sometimes splitting up the method. You could write an extension method targeting HostApplicationBuilder, but we don't recommend that because it's sealed.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Program.cs [hunk 1]
```diff
@@ -1,10 +1,7 @@
-using Company.Application1;
```
Comment (halter73) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Worker-CSharp/Program.cs:1]: Isn't `Worker` still in this namespace?

**Comment by ReubenBond** [comment]
> It probably also break orleans right @ReubenBond ?

It breaks 3.x, yeah. In 4.x, we have a method to add/configure Orleans via `IServiceCollection` instead of only `IHostBuilder`

**Comment by davidfowl** [comment]
We may need to hold on this one, [a quick browse](https://grep.app/search?q=this%20IHostBuilder) shows that some important 3rd party packages have built extensions on IHostBuilder.

**Comment by davidfowl** [comment]
Yep, I'm convinced this is a bad decision for 7.0, nothing will work with this new builder so we should give people time to adopt it before making everyone revert to the old builder because it doesn't work with their package.

**Comment by halter73** [comment]
I doubt migrating from the generic host to HostApplicationBuilder will have the same wow factor as migrating from the generic web host to WebApplicationBuilder. Not having to use callbacks is nice, but I don't expect many people to switch just for that.

Without using it in the templates and docs, I don't see many people migrating. I think it's okay if most people continue to use HostBuilder though. HostApplicationBuilder has allowed us to clean up WebApplicationBuilder and fix a lot of bugs, and the API is still public for those that prefer less callbacks.

**Pull request closed without merging**